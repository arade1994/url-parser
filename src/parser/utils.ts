const URL_REGEX = /(https?:\/\/[^\s\]]+|www\.[^\s\]]+)/gi;

//**
/**
 * Extract the last URL from a text
 * @param text - The text to extract the last URL from
 * @returns The last URL in the text, or null if no URL is found
 */
export function extractLastUrlFromText(text: string): string | null {
  let lastUrl: string | null = null;
  let matchedUrls: RegExpExecArray | null = null;
  while ((matchedUrls = URL_REGEX.exec(text)) !== null) lastUrl = matchedUrls[0];

  return lastUrl;
}
