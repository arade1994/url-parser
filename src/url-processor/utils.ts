import * as crypto from 'node:crypto';
import type { IUrlInfo } from './types';

/**
 * Formats the URL to be a valid URL
 * @param url - The URL to format
 * @returns The formatted URL
 */
export function formatUrl(url: string): string {
  return /^https?:\/\//i.test(url) ? url : `https://${url}`;
}

/**
 * Fetches the body with a GET request to the URL
 * @param url - The URL to fetch the body of
 * @returns The body of the URL as a string
 */
export async function fetchUrlBody(url: string): Promise<string> {
  let response;
  try {
    response = await fetch(url);
  } catch (error: any) {
    throw new Error(`Network error: ${error.message}`);
  }

  if (!response.ok) throw new Error(`HTTP ${response.status}`);
  return await response.text();
}

/**
 * Extracts the title and email from the HTML
 * @param url - The URL to extract the title and email from
 * @param html - The HTML of the URL
 * @returns The parsed URL info
 */
export function extractUrlTitleAndEmail(url: string, html: string): IUrlInfo {
  const info: IUrlInfo = { url };
  const title = /<title[^>]*>([\s\S]*?)<\/title>/i.exec(html)?.[1]?.trim();
  if (title) info.title = title;

  const emailMatch = /[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}/i.exec(html)?.[0];
  if (emailMatch) info.email = emailMatch;

  return info;
}

/**
 * Hashes the email
 * @param email - The email to hash
 * @param secret - The secret to use for the hash
 * @returns The hashed email
 */
export function hashEmail(email: string, secret: string): string {
  return crypto.createHmac('sha256', secret).update(email, 'utf8').digest('hex');
}
