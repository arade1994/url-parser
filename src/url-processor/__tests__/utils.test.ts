import { formatUrl, extractUrlTitleAndEmail, hashEmail, fetchUrlBody } from '../utils';

describe('formatUrl()', () => {
  test('adds https:// if missing', () => {
    expect(formatUrl('www.google.com')).toBe('https://www.google.com');
  });

  test('keeps existing http:// prefix', () => {
    expect(formatUrl('http://example.com')).toBe('http://example.com');
  });

  test('keeps existing https:// prefix', () => {
    expect(formatUrl('https://example.com')).toBe('https://example.com');
  });
});

describe('fetchUrlBody()', () => {
  test('fetches url body', async () => {
    expect(await fetchUrlBody('https://example.com')).toMatchSnapshot();
  });
});

describe('extractUrlTitleAndEmail()', () => {
  test("doesn't extract title and email if no title or email is found", () => {
    const html = '<html><head><body><p>This is an example website.</p></body></html>';
    expect(extractUrlTitleAndEmail('https://example.com', html)).toEqual({
      url: 'https://example.com',
      title: undefined,
      email: undefined,
    });
  });

  test('extracts title with no email', () => {
    const html =
      '<html><head><title>Example Domain</title></head><body><p>This is an example website.</p></body></html>';
    expect(extractUrlTitleAndEmail('https://example.com', html)).toEqual({
      url: 'https://example.com',
      title: 'Example Domain',
      email: undefined,
    });
  });

  test('extracts email with no title', () => {
    const html = '<html><head></head><body><p>example@example.com</p></body></html>';
    expect(extractUrlTitleAndEmail('https://example.com', html)).toEqual({
      url: 'https://example.com',
      title: undefined,
      email: 'example@example.com',
    });
  });

  test('extracts title and email', () => {
    const html =
      '<html><head><title>Example Domain</title></head><body><p>This is an example website.</p><p>example@example.com</p></body></html>';
    expect(extractUrlTitleAndEmail('https://example.com', html)).toEqual({
      url: 'https://example.com',
      title: 'Example Domain',
      email: 'example@example.com',
    });
  });
});

describe('hashEmail()', () => {
  test('hashes email', () => {
    expect(hashEmail('example@example.com', 'secret')).toBe(
      '0bb2275df8909502beb1a257c30d55d734d24e1172def381244b602184816d8f',
    );
  });
});
