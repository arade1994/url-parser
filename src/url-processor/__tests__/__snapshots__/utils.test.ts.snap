// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetchUrlBody() > fetches url body 1`] = `"<html><title>Mock Page</title><body><p>Email: mock@example.com</p></body></html>"`;
